sql/core/src/main/scala/org/apache/spark/sql/GroupedData.scala

```diff
@@ -39,13 +39,14 @@ class GroupedData protected[sql](df: DataFrameImpl, groupingExprs: Seq[Expressio
       df.sqlContext, Aggregate(groupingExprs, namedGroupingExprs ++ aggExprs, df.logicalPlan))
   }
 
-  private[this] def aggregateNumericColumns(f: Expression => Expression): Seq[NamedExpression] = {
-    df.numericColumns.map { c =>
-      val a = f(c)
-      Alias(a, a.toString)()
-    }
+  private[this] def aggregateNumericColumns(colNames: String*)
```

Comment:
can you change this to

```scala
  private[this] def aggregateNumericColumns(colNames: String*)(f: Expression => Expression)
    : Seq[NamedExpression] = {

    val columnExprs = if (colNames.isEmpty) {
      // No columns specified. Use all numeric columns.
      df.numericColumns
    } else {
      // Make sure all specified columns are numeric
      colNames.map { colName =>
        val namedExpr = df.resolve(colName)
        if (!namedExpr.dataType.isInstanceOf[NumericType]) {
          throw new AnalysisException(
            s""""$colName" is not a numeric column. """ +
            "Aggregation function can only be performed on a numeric column.")
        }
        namedExpr
      }
    }
    columnExprs.map { c =>
      val a = f(c)
      Alias(a, a.toString)()
    }
  }
```

and just remove the old one? There is actually a bug in the old one, because it doesn't use a resolver to compare column names. On top of that, it is not uncommon tables have thousands of columns. You have a operation that is actually O(n^2)...